### PR TITLE
Serialization for per channel qtensor

### PR DESCRIPTION
**Pull request by dzhulgakov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26243 Fix _empty_per_channel_affine_quantized to be less hacky
* **#26339 Serialization for per channel qtensor**

Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

Differential Revision: [D17443222](https://our.internmc.facebook.com/intern/diff/D17443222)

### CODE REVIEW

File: torch/csrc/jit/pickler.cpp [hunk 1]
```diff
@@ -403,8 +404,41 @@ void Pickler::pushLiteralTensor(const IValue& ivalue) {
   push<PickleOpCode>(PickleOpCode::TUPLE);
 
   if (quantized) {
-    pushDouble(tensor.q_scale());
-    pushInt(tensor.q_zero_point());
+    push<PickleOpCode>(PickleOpCode::MARK);
+    pushGlobal("torch", toString(tensor.qscheme()));
+    // tuple of (qscheme, scale, zp) or (qscheme, scales, zps, axis)
```
Comment (driazati) [inline torch/csrc/jit/pickler.cpp]: Why are there two different schemas here? Is it possible to always do it as `(qscheme, scales, zps, axis)`? Specializing is going to make the (de)serialization logic much more complicated

File: torch/csrc/jit/pickler.cpp [hunk 2]
```diff
@@ -403,8 +404,41 @@ void Pickler::pushLiteralTensor(const IValue& ivalue) {
...
+    switch (tensor.qscheme()) {
+      case at::kPerTensorAffine:
+        pushDouble(tensor.q_scale());
+        pushInt(tensor.q_zero_point());
+        break;
+      case at::kPerChannelAffine: {
+        const auto* quantizer = static_cast<at::PerChannelAffineQuantizer*>(
+            tensor.quantizer().get());
+        push<PickleOpCode>(PickleOpCode::MARK);
```
Comment (driazati) [inline torch/csrc/jit/pickler.cpp]: You can shorten all of these by creating a `c10::tuple` and using `pushIValue`

File: torch/csrc/jit/pickler.cpp [hunk 3]
```diff
@@ -812,18 +846,53 @@ PickleOpCode Unpickler::readInstruction() {
...
+                auto* zero_points_ptr = zero_points.data_ptr<int64_t>();
+                for (int i = 0; i < (int)zero_points_tuple.size(); ++i) {
+                  zero_points_ptr[i] = zero_points_tuple[i].toInt();
+                }
+                std::vector<int64_t> axis;
+                for (const auto& x : qparams.at(3).toTuple()->elements()) {
+                  axis.push_back(x.toInt());
+                }
+                result = _empty_per_channel_affine_quantized(
+                    {0}, scales, zero_points, axis, storage_tensor.options());
```
Comment (driazati) [inline torch/csrc/jit/pickler.cpp]: Why are all the parameters for `_empty_per_channel_affine_quantized` `Tensor`s instead of a list?

**Review by driazati — COMMENTED** [review]
I'm generally apprehensive on the serialization changes, there's a lot of special casing / deeply paths in the tensor serialization and it's going to be hard to ensure there's no forward/backward compatibility problems.

**Comment by driazati** [comment]
As for tests that compare, you can try to save with the pickler (e.g. use `torch.save` in a `@script` function) and load it in eager, and vice versa. This test does that for regular tensors:


https://github.com/pytorch/pytorch/blob/06c69ad8ede8aa90a035da6405a3267fa5efe915/test/test_jit.py#L14797

File: test/common_utils.py [hunk 4]
```diff
@@ -722,11 +722,21 @@ def assertTensorsEqual(a, b):
             elif x.is_quantized and y.is_quantized:
```
Comment (jerryzh168) [inline test/common_utils.py]: nit: we can remove `and y.is_quantized` here

File: test/test_quantized_tensor.py [hunk 5]
```diff
@@ -208,13 +208,29 @@ def test_qtensor_per_channel_permute(self):
...
+                self.assertEqual(qrv, qrv2)
+                self.assertEqual(qr2.storage().data_ptr(), qrv2.storage().data_ptr())
+
+    def test_qtensor_per_channel_load_save(self):
+        r = torch.rand(20, 10, dtype=torch.float) * 4 - 2
+        scales = torch.rand(10) * 0.02 + 0.01
+        zero_points = torch.round(torch.rand(10) * 20 + 1).to(torch.long)
+        # quint32 is not supported yet
+        for dtype in [torch.quint8, torch.qint8]:
+            qr = torch.quantize_linear_per_channel(r, scales, zero_points, [1], dtype)
```
Comment (jerryzh168) [inline test/test_quantized_tensor.py]: what is the behavior for slicing(view) on per channel quantized tensor right now?

**Review by jerryzh168 — APPROVED** [review]
checked code except `jit/pickler.cpp`.

File: torch/csrc/jit/pickler.cpp (hunk 3 above)
Comment (jerryzh168) [reply torch/csrc/jit/pickler.cpp]: because `scales` is float array, we don't have float array and we don't have plans to support it

File: torch/csrc/jit/pickler.cpp (hunk 1 above)
Comment (dzhulgakov) [reply torch/csrc/jit/pickler.cpp]: It depends on quantizer (which can be per-tensor or per-channel) - the per-tensor one doesn't have axis. I can make axis to be None in the other case, but not sure it makes it cleaner. Specialization still has to be there as the scale vs scales is int vs tuple of int

File: torch/csrc/jit/pickler.cpp (hunk 3 above)
Comment (dzhulgakov) [reply torch/csrc/jit/pickler.cpp]: also, very theoretically, we might have multi-axis quantization in which case scales would be a multidimensional tensor. It's good to have a support for it in api, but it's not likely to appear soon. (note that in serialiation it's still fine to fold the scales to a tuple as the shape is deducible from the original tensor.

File: test/test_quantized_tensor.py (hunk 5 above)
Comment (dzhulgakov) [reply test/test_quantized_tensor.py]: any slicing / set strides operation today basically errors out with nice message of "can't do it".  We could support a subset of these operations, but I doubt it's worth it at this point.

Comment (raghuramank100) [reply test/test_quantized_tensor.py]: Is this true for permute also? We could conceptually support it, but again doesn't  seem like it is really needed

File: torch/tensor.py [hunk 6]
```diff
@@ -41,14 +41,27 @@ def __reduce_ex__(self, proto):
         # See Note [Don't serialize hooks]
         torch.utils.hooks.warn_if_has_hooks(self)
         if self.is_quantized:
+            if self.qscheme() == torch.per_tensor_affine:
```
Comment (raghuramank100) [inline torch/tensor.py]: We still support torch.per_tensor_symmetric and torch.per_channel_symmetric as valid qschemes. They are used only by the observer currently. We should add checks for these too whenever we have a qscheme based check.

File: torch/csrc/jit/pickler.cpp [hunk 7]
```diff
@@ -786,18 +808,46 @@ PickleOpCode Unpickler::readInstruction() {
...
+            const auto& qparams = qparams_tuple->elements();
+            auto qscheme = static_cast<at::QScheme>(qparams.at(0).toInt());
+            switch (qscheme) {
+              case at::kPerTensorAffine: {
+                double q_scale = qparams.at(1).toDouble();
+                int64_t q_zero_point = qparams.at(2).toInt();
+                result = at::_empty_affine_quantized(
+                    {0}, storage_tensor.options(), q_scale, q_zero_point);
+              } break;
+              case at::kPerChannelAffine: {
```
Comment (raghuramank100) [inline torch/csrc/jit/pickler.cpp]: We should also handle kPerChannelSymmetric and kPerTensorSymmetric here

File: test/test_jit.py [hunk 8]
```diff
@@ -3297,23 +3297,29 @@ def test_print_op_module(self):
...
             def forward(self):
                 return self.x
 
-        model = SimpleQTensor()
-        buffer = io.BytesIO()
-        torch.jit.save(model, buffer)
-        buffer.seek(0)
-        model_loaded = torch.jit.load(buffer)
-        self.assertEqual(model_loaded(), model())
+        for per_channel in [False, True]:
```
Comment (raghuramank100) [inline test/test_jit.py]: Can we also sweep over qschemes in this test?

File: test/test_jit.py [hunk 9]
```diff
@@ -14841,6 +14847,34 @@ def forward(self, x):
...
+                def __init__(self):
+                    super(M, self).__init__()
+                    self.fname = fname
+
+                @torch.jit.script_method
+                def forward(self, x, y):
+                    torch.save((x, y), self.fname)
+                    return y
+
+            q = torch.quantize_linear(
```
Comment (raghuramank100) [inline test/test_jit.py]: Suggest also sweeping over qschemes

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please add support for symmetric qschemes.

File: test/test_jit.py (hunk 8 above)
Comment (dzhulgakov) [reply test/test_jit.py]: @raghuramank100 - it's already sweeping among the two qschemes we support. We don't currently pass qscheme to any API and there are two calls quantize_linear and quantize_linear_per_channel to create tensor, so I don't think we can create symmetric tensor today

File: test/test_jit.py (hunk 9 above)
Comment (dzhulgakov) [reply test/test_jit.py]: it does sweep on the two we support (See `qc` below)

File: test/test_quantized_tensor.py (hunk 5 above)
Comment (dzhulgakov) [reply test/test_quantized_tensor.py]: yes, `permute` is implemented as strides setting, so it gives the same error

File: torch/csrc/jit/pickler.cpp (hunk 7 above)
Comment (dzhulgakov) [reply torch/csrc/jit/pickler.cpp]: again, we have only two types of quantizers present today

File: torch/tensor.py (hunk 6 above)
Comment (dzhulgakov) [reply torch/tensor.py]: this code will never be hit by observers - you can't create a tensor with that quantizer

**Review by raghuramank100 — APPROVED** [review]
